src/http/delta.rs

```diff
@@ -0,0 +1,327 @@
+) -> Response<Body> {
+    Response::builder()
+    .header("Content-Type", "application/json")
+    .body(Body::wrap_stream(stream::iter(
+        SnapshotStream::new(session, to_serial, snapshot)
+        .map(Result::<_, Infallible>::Ok)
+    )))
+    .unwrap()
+}
+
+
+//------------ DeltaStream ---------------------------------------------------
+
+/// An iterator as the foundation for streaming a delta.
+///
+/// The iterator produces segments of roughly 64k size. This can be converted
+/// into a async stream and then used with Tokio’s `Body::wrap_stream`.
+struct DeltaStream {
+    /// The header of the output.
+    ///
+    /// This is set to some when a new value is created and then taken out on
+    /// the very first iteration.
+    header: Option<Vec<u8>>,
+
+    /// The delta we work on.
+    ///
+    /// This is set to `None` to fuse the iterator.
+    delta: Option<Arc<PayloadDelta>>,
+
+    /// The position of the next item in the delta.
+    ///
+    /// If this is `Ok(_)`, we are working with the announced origins. If this
+    /// is `Err(_)` we are working with the withdrawn origins.
+    pos: Result<usize, usize>,
+}
+
+impl DeltaStream {
+    /// Creats a new delta stream.
+    fn new(
+        session: u64, from_serial: Serial, to_serial: Serial,
+        delta: Arc<PayloadDelta>
+    ) -> Self {
+        let mut vec = Vec::new();
+        Self::append_header(&mut vec, session, from_serial, to_serial);
+        DeltaStream {
+            header: Some(vec),
+            delta: Some(delta),
+            pos: Ok(0),
+        }
+    }
+
+    /// Appends the delta header to the provided vec.
+    fn append_header(
+        vec: &mut Vec<u8>,
+        session: u64, from_serial: Serial, to_serial: Serial,
+    ) {
+        write!(vec, "\
+            {{
+            \n  \"reset\": false,\
+            \n  \"session\": \"{}\",\
+            \n  \"serial\": {},\
+            \n  \"fromSerial\": {},\
+            \n  \"announced\": [",
+            session, to_serial, from_serial
+        ).unwrap()
+    }
+
+    /// Appends the separator between announced and withdrawn to the vec.
+    fn append_separator(
+        vec: &mut Vec<u8>,
+    ) {
+        write!(vec, "\
+            \n  ],\
+            \n  \"withdrawn\": [",
+        ).unwrap()
+    }
+
+    /// Appends an origin to the vec.
+    fn append_origin(
+        vec: &mut Vec<u8>,
+        origin: RouteOrigin,
+        first: bool
+    ) {
+        if !first {
+            vec.push(b',')
+        }
+        write!(vec, "\
+            \n    {{ \"asn\": \"{} \", \"prefix\": \"{}/{}\", \
+                     \"maxLength\": {} }}",
+            origin.as_id(),
+            origin.address(), origin.address_length(),
+            origin.max_length()
+        ).unwrap()
+    }
+
+    /// Appends the footer to the vec.
+    fn append_footer(vec: &mut Vec<u8>) {
+        vec.extend_from_slice(b"\n  ]\n}\n");
+    }
+}
+
```

Comment:
Any particular reason for not using Serde here?

Reply:
Yes: This implementation allows to stream the result in blocks to save memory. I don’t think this can be done with serde.